### ISSUE TITLE
Guide to integrate mlflow spark automatic logging using java API

### ISSUE THREAD

**Issue by sreerama-naga** [mlflow/mlflow#5506]
Hi,

Looking for a guide or document which explains about  how to integrate spark automatic logging using java API.
Understood that the _mlflow-spark_ library is the one to use for integration.
 
Python API documentation is available here https://www.mlflow.org/docs/latest/tracking.html#spark.
Looking for similar type of guide/documentation  with steps for java API. 

Precisely,
1) What are the configurations to do.
2) what are the different API's (example: methods) to be invoked
3) what are all the details can be logged automatically
4) Does the mlflow-spark jar works well with scala 2.13 version. If not, is it possible to make it to work with scala 2.13  
5) Does mlflow-spark jar works well with spark 3.2.0 ?  ([pom](https://repo1.maven.org/maven2/org/mlflow/mlflow-spark/1.24.0/mlflow-spark-1.24.0.pom) says, it used scala 2.12 and spark 3.0-preview. )

Thank you in advance.

**Comment by BenWilson2** [comment]
Hi @sreerama-naga , due to the relatively low utilization of the JavaAPI, we haven't spent much time on ensuring complete feature parity between the primary Python APIs and the Java APIs. So, for the missing functionality, such as autologging, I'm afraid that we just don't have that implementation prioritized, opting instead to work on features within the Python API. 
That being said, we would certainly not turn down a submission for a solution (although the complexity of implementing this will likely be extreme). 
To answer each of your questions:
1. None. You can't do it with the current version.
2. There are none. Since autologging relies on patching of invoked methods during runtime, any Java or Scala models that are running would need to have runtime inspection implemented in those languages to enable features such as this. While entirely possible, it's just not something that is even remotely close to the top of our list of priorities for MLflow.
3. In the PythonAPI, for PySpark models, auto-logging is fully featured (with the exception of the soon-to-be-implemented signature and example auto-logging, which will hopefully be released in either the next release version or the one after that). There simply is no similar functionality (metrics, params, artifacts, etc. all must be logged manually with calls to the JavaAPI wrapper around the Python APIs).
4. The current supported (as of MLflow 1.24.0) version of scala is 2.12.6 as per https://mvnrepository.com/artifact/org.mlflow/mlflow-spark/1.24.0 . This isn't to say that it won't work with scala 2.13 (you can always fork MLflow, update the POM, and use maven to build a version with your required scala runtime and see if it works); we simply haven't tested it with 2.13. Yet. We'll be taking a look at this since Spark 3.2.0 introduced support for 2.13 in November. 
5. Spark 3.2.0 and 3.2.1 are built on scala 2.12 and scala 2.13: https://mvnrepository.com/artifact/org.apache.spark/spark-core. It'll work fine! :) 

Hope that helps!

**Comment by sreerama-naga** [comment]
@BenWilson2 thank so much for the details. It certainly helps.

One more question:
Does save or log model functionality is still not available with Java API? means the answer [here](
https://github.com/mlflow/mlflow/issues/1595) is still valid?

Alternate option I see, to log_model is treat the models as artifacts and log as artifact by taking advantage of log_artifact(s) functions. But I am not sure how far logging as artifact compete with log_model in behavior wise.

**Comment by amesar** [comment]
The Java/Scala client is a limited subset of the full-fledged Python implementation. Much of the rich Python functionality (flavors, projects) is implemented on the client side. Due to low demand, this has not been implemented for Java/Scala. For Java/Scala there is no concept of ML "flavors" (sklearn, keras, etc.) so you have to log your model as a barebones artifact.

**Comment by sreerama-naga** [comment]
@amesar thank you.
